matmul.hip

```diff
@@ -2788,6 +2791,341 @@ class MmtKernel_StreamKTree_256t_MSxNS_amdgcn_mfma_f32_16x16x16f16_shared_Kx2
+    // Accumulator VGPRs.
+    acc_array_t acc = {{0}};
+    int tid = threadIdx.x;
+    constexpr int A_tile_size_in_vec8 = MS * 16 * 4;
+    constexpr int B_tile_size_in_vec8 = NS * 16 * 4;
+    const float16x8_t *A_global =
+        static_cast<const float16x8_t *>(A_data) +
+        (m_outer * K_outer + k_iter_start) * A_tile_size_in_vec8;
+    const float16x8_t *B_global =
+        static_cast<const float16x8_t *>(B_data) +
+        (n_outer * K_outer + k_iter_start) * B_tile_size_in_vec8;
+    const float16x8_t *A_global_ptr = A_global + tid;
+    const float16x8_t *B_global_ptr = B_global + tid;
+
+    // Shared memory buffers for tiles of A and B matrices.
+    __shared__ float16x8_t A_shared[A_tile_size_in_vec8];
+    __shared__ float16x8_t B_shared[B_tile_size_in_vec8];
+
+    // Inner loop on K dimenion.
+    for (int k_iter = k_iter_start; k_iter < k_iter_end; ++k_iter) {
+      // Load data from global to shared memory.
+      for (int i = 0; i < A_tile_size_in_vec8; i += 256) {
+        A_shared[i + tid] = A_global_ptr[i];
+      }
+      for (int j = 0; j < B_tile_size_in_vec8; j += 256) {
+        B_shared[j + tid] = B_global_ptr[j];
+      }
+      A_global_ptr += A_tile_size_in_vec8;
+      B_global_ptr += B_tile_size_in_vec8;
+      __syncthreads();
+      // Perform MFMA arithmetic on data in shared memory.
+      for (int i = 0; i < MS; ++i) {
+        for (int j = 0; j < NS / 4; ++j) {
+          float16x8_t a = A_shared[64 * i + (tid % 64)];
+          float16x8_t b = B_shared[256 * j + tid];
+          float16x4_t a0 = (float16x4_t){a[0], a[1], a[2], a[3]};
+          float16x4_t a1 = (float16x4_t){a[4], a[5], a[6], a[7]};
+          float16x4_t b0 = (float16x4_t){b[0], b[1], b[2], b[3]};
+          float16x4_t b1 = (float16x4_t){b[4], b[5], b[6], b[7]};
+          floatx4_t acc0 = acc[i * (NS / 4) + j];
+          acc0 = __builtin_amdgcn_mfma_f32_16x16x16f16(a0, b0, acc0, 0, 0, 0);
+          acc0 = __builtin_amdgcn_mfma_f32_16x16x16f16(a1, b1, acc0, 0, 0, 0);
+          acc[i * (NS / 4) + j] = acc0;
+        }
+      }
+      __syncthreads();
+    }
+    return acc;
+  }
+  __device__ static uint32_t workgroup_atomic_or(atomic_t *atomic,
+                                                 uint32_t value) {
+    __shared__ int workgroup_value;
+    int tid = threadIdx.x;
+    if (tid == 0) {
+      // Thread0 fetches and increments the device-wide atomic.
+      // This is a relaxed atomic, which is fine, as we provide the memory
+      // ordering separately below with __builtin_amdgcn_fence, which usefully
+      // allows us to specify the memory scope.
+      int thread0_value = atomicOr(atomic, value);
+      // Thread0 stores the value to the workgroup-wide shared location.
+      workgroup_value = thread0_value;
+      // Memory scope is "agent" because we need other workgroups to see this.
+      __builtin_amdgcn_fence(__ATOMIC_RELEASE, "agent");
```

Comment:
I thought that agent is the whole gpu? Is there a finer level of granularity?

Reply:
It is the whole logical device, intentionally as this is inter-workgroup synchronization.  The only potential finer level of granularity here would be the XCD, however that would require adjusting the stream-k distribution to ensure that each tile is entirely processed within one XCD.  That would not be too hard, and potentially a nice speed-up, and arguably we don't care about the extreme cases where we would attempt splitting a tile across XCDs. So, +1 to that but:  (1) as a follow-up, as I am currently dealing with far worse performance issues already in the data-parallel case; (2) if anyone knows how to specify XCD-scope atomics, let me know. I can't see it on https://llvm.org/docs/AMDGPUUsage.html .

Reply:
Oh right, it can't be just the workgroup scope